#2726 Legend occupies too much space in reports screen

Patrykb0802 opened the pull request:
This pull requests concerns: #2726 

Added:
- In ReportChartCreator.java, a variable that adds to height (of legend and consolidated chart) for every point added to report

Patrykb0802 committed:
#2726 Legend occupies too much space in reports screen

Added:
- In ReportChartCreator.java, a variable that adds to height (of legend and consolidated chart) for every point added to report

Patrykb0802 committed:
#2726 Legend occupies too much space in reports screen

Added:
- Removed unused imports

Limraj commented on src/com/serotonin/mango/vo/report/ReportChartCreator.java:
```diff
@@ -58,6 +58,7 @@ public class ReportChartCreator {
      */
     private static final int IMAGE_WIDTH = 930;
     private static final int IMAGE_HEIGHT = 400;
+    private static final int IMAGE_HEIGHT_PX_INCREMENT_PER_POINT = 17;
```
Word Increment does not make sense here, it is the number of pixels per point label.
I propose a name:
IMAGE_POINT_LABEL_HEIGHT_PIXELS

Patrykb0802 committed:
#2726 Legend occupies too much space in reports screen

- Changed variable name

Patrykb0802 committed:
- Changed name of variable responsible for legend height

Limraj commented on src/com/serotonin/mango/vo/report/ReportChartCreator.java:
```diff
@@ -145,7 +146,7 @@ else if (pointStat.getDiscreteTimeSeries() != null)
             }
 
-            imageData = ImageChartUtils.getChartData(ptsc, true, IMAGE_WIDTH, IMAGE_HEIGHT);
+            imageData = ImageChartUtils.getChartData(ptsc, true, IMAGE_WIDTH, (IMAGE_HEIGHT+(pointStatistics.size()* IMAGE_HEIGHT_FOR_DATA_POINT_NAME_IN_LEGEND_PIXELS)));
```
Here you need to move these calculations to a separate method

Patrykb0802 committed:
Code refactor

Added:
- calculateHeightConsolidatedChart method to calculate height needed for consolidated graph in ImageChartUtils.java
- CalculateConsolidatedChartHeightTest.java for testing calculateHeightConsolidatedChart method

Changed:
- In ReportInstancePointDAO.java how names of data source and data point are passed, now they are truncated before passing them to ReportChartCreator

Patrykb0802 committed:
Code refactor

Added:
- ReplaceMacroStringUtilsWithProdParametersTest.java for testing RepalceMacro method

Changed:
- insert method in ReportInstancePointDAO.java so now it takes less parameters (uses only pointInfo now for all necessary data point information)
- truncate method to abbreviateMiddle for truncating data source and data point names so they fit report chart

Limraj commented on src/com/serotonin/mango/vo/report/ReportChartCreator.java:
```diff
@@ -144,8 +156,11 @@ else if (pointStat.getDiscreteTimeSeries() != null)
-
-            imageData = ImageChartUtils.getChartData(ptsc, true, IMAGE_WIDTH, IMAGE_HEIGHT);
+            try{
+                imageData = ImageChartUtils.getChartData(ptsc, true, IMAGE_WIDTH, ImageChartUtils.calculateHeightConsolidatedChart(pointStatistics, IMAGE_HEIGHT, IMAGE_HEIGHT_FOR_DATA_POINT_NAME_IN_LEGEND_PIXELS, CHARACTERS_PER_LINE_IN_GRAPH_LEGEND));
```
int consolidatedChartHeight = ImageChartUtils.calculateHeightConsolidatedChart(pointStatistics, IMAGE_HEIGHT, IMAGE_HEIGHT_FOR_DATA_POINT_NAME_IN_LEGEND_PIXELS, CHARACTERS_PER_LINE_IN_GRAPH_LEGEND)
then:
imageData = ImageChartUtils.getChartData(ptsc, true, IMAGE_WIDTH, consolidatedChartHeight);

Limraj commented on src/com/serotonin/mango/vo/report/ReportChartCreator.java:
```diff
@@ -58,6 +58,10 @@ public class ReportChartCreator {
     private static final int IMAGE_WIDTH = 930;
     private static final int IMAGE_HEIGHT = 400;
+	private static final int IMAGE_HEIGHT_FOR_DATA_POINT_NAME_IN_LEGEND_PIXELS = 20;
+    private static final int CHARACTERS_PER_LINE_IN_GRAPH_LEGEND = 138; //this value refers to IMAGE_WIDTH equal to 930
```
NUMBER_CHARACTERS_PER_LINE_IN_CHART_LEGEND

Limraj commented on src/com/serotonin/mango/vo/report/ReportChartCreator.java:
```diff
@@ -58,6 +58,10 @@ public class ReportChartCreator {
     private static final int IMAGE_HEIGHT = 400;
+	private static final int IMAGE_HEIGHT_FOR_DATA_POINT_NAME_IN_LEGEND_PIXELS = 20;
+    private static final int CHARACTERS_PER_LINE_IN_GRAPH_LEGEND = 138; //this value refers to IMAGE_WIDTH equal to 930
+	private static final int DATA_SOURCE_NAME_LENGTH_FOR_GRAPH = 35; // leaving 3 characters of space for " - " between data source name and data point name in extended name
```
DATA_SOURCE_NAME_LENGTH_FOR_CHART

Limraj commented on src/com/serotonin/mango/vo/report/ReportChartCreator.java:
```diff
@@ -58,6 +58,10 @@ public class ReportChartCreator {
+	private static final int IMAGE_HEIGHT_FOR_DATA_POINT_NAME_IN_LEGEND_PIXELS = 20;
+    private static final int CHARACTERS_PER_LINE_IN_GRAPH_LEGEND = 138; //this value refers to IMAGE_WIDTH equal to 930
+	private static final int DATA_SOURCE_NAME_LENGTH_FOR_GRAPH = 35; // leaving 3 characters of space for " - " between data source name and data point name in extended name
+	private static final int DATA_POINT_NAME_LENGTH_FOR_GRAPH = 100;
```
DATA_POINT_NAME_LENGTH_FOR_CHART

Limraj commented on src/org/scada_lts/serorepl/utils/StringUtils.java:
```diff
@@ -133,20 +133,15 @@ public static String trimWhitespace(String s){
         return s.trim();
     }
 
-    public static String truncate(String s, int length){
```
We are not removing this method, we have created another one to shorten names. If this method is only used in tests, and it behaves differently than the same method from the Apache common library, so its operation is not intuitive, then I would set the `@Deprecated` annotation on it.

Limraj commented on test/com/serotonin/mango/vo/report/CalculateConsolidatedChartHeightTest.java:
```diff
@@ -0,0 +1,401 @@
+	private final int imageHeight;
+	private final int charAmountPerLine;
+	private final boolean expectException;
+	static List<String> pointNames = Arrays.asList(
```
Too much data, tests have to be understandable to humans, so we try to limit the complexity of cases to the necessary minimum, of course there may be one few cases with a large amount of data, but they should not dominate. The idea is to check the simplest possible cases that test well the behavior in a real application. The test should be written in such a way that if it fails, the person immediately knows what the reason is, what needs to be corrected, and in what scenario it was broken.

Limraj commented on test/com/serotonin/mango/vo/report/CalculateConsolidatedChartHeightTest.java:
```diff
@@ -0,0 +1,401 @@
+	}
+
+	@Test
+	public void when_calculateConsolidatedChartHeight() {
```
To make the tests more readable, I recommend moving test cases expecting an exception to a separate class and there using the construction intended for this purpose from the commented test method, i.e. `@Test(expected = IllegalArgumentException.class)`

Limraj commented on test/org/scada_lts/serorepl/utils/StringUtilsTest.java:
```diff
@@ -195,19 +195,19 @@ public void replaceMacros() throws Exception {
     public void truncate2Parameters() throws Exception {
         assertEquals(StringUtils.truncate("abcd", 3, "XXX"),
-                org.scada_lts.serorepl.utils.StringUtils.truncate("abcd", 3, "XXX")
+                org.scada_lts.serorepl.utils.StringUtils.abbreviateMiddle("abcd", 3, "XXX")
```
Here we go back to the old method, but we set Deprecated on the test class.

Limraj commented on src/com/serotonin/mango/vo/report/ImageChartUtils.java:
```diff
@@ -210,4 +211,27 @@ public static void addSecond(TimeSeries timeSeries, long time, Number value) {
         }
     }
+
+	public static int calculateHeightConsolidatedChart(List<ReportChartCreator.PointStatistics> pointStatistics, int imageHeight, int imageHeightForDataPointNameInLegend, int charAmountPerLine) {
```
charAmountPerLine -> charsNumberPerLine

Patrykb0802 committed:
Code refactor

- Moved tests for IllegalArgumentException to CalculateConsolidatedChartHeightTestException.java
- Small changes in variable names
- Deprecated truncate method in StringUtils.java and truncate method tests in StringUtilsTest.java

Limraj commented on src/com/serotonin/mango/vo/report/ReportChartCreator.java:
```diff
@@ -58,6 +58,10 @@ public class ReportChartCreator {
     private static final int IMAGE_WIDTH = 930;
     private static final int IMAGE_HEIGHT = 400;
+	private static final int IMAGE_HEIGHT_FOR_DATA_POINT_NAME_IN_LEGEND_PIXELS = 20;
+    private static final int NUMBER_CHARACTERS_PER_LINE_IN_CHART_LEGEND = 138; //this value refers to IMAGE_WIDTH equal to 930
```
We try to avoid comments, the variable name should say what it is, comment is unnecessary. And it does not refer to this parameter at all, the legend parameter does not apply to the entire generated image with a chart;

CHARACTERS_PER_LINE_IN_CHART_LEGEND_NUMBER

Limraj commented on src/com/serotonin/mango/vo/report/ReportChartCreator.java:
```diff
@@ -58,6 +58,10 @@ public class ReportChartCreator {
      */
     private static final int IMAGE_WIDTH = 930;
     private static final int IMAGE_HEIGHT = 400;
+	private static final int IMAGE_HEIGHT_FOR_DATA_POINT_NAME_IN_LEGEND_PIXELS = 20;
```
you need to correct these spaces/tabs. 
This parameter should be associated with the legend and not with the height of the entire image. This is a label, not a name, you can imagine the height and width of the label, and on average the name of the point, what would it mean?
IMAGE_POINT_LABEL_HEIGHT_IN_LEGEND_PIXELS

Limraj commented on src/org/scada_lts/serorepl/utils/StringUtils.java:
```diff
@@ -133,20 +133,33 @@ public static String trimWhitespace(String s){
-        if (truncateSuffix != null){
-            return s.substring(0,length) + truncateSuffix;
+	@Deprecated
+	public static String truncate(String s, int length) {
```
Only what has been changed in the code should be visible, not the entire block.

Patrykb0802 committed:
Code refactor

- Changed test's dataset names
- Small changes in code appearance

Patrykb0802 committed:
indent change

Patrykb0802 committed:
indent correction,
Added more test cases for CalculateConsolidatedChartHeightTest.java

Patrykb0802 committed:
indent correction

Patrykb0802 committed:
indent correction

Patrykb0802 committed:
indent correction

Patrykb0802 committed:
Data for the tests has been changed
Added condition for calculateHeightConsolidatedChart method, so now it handles borderline cases better

Patrykb0802 committed:
Added testName to each dataset used for testing in CalculateConsolidatedChartHeightTest.java for better understanding which test fails if such thing happens

Patrykb0802 committed:
Code reformatting, getting rid of unnecessary test datasets
Removed createPointStatistics generator in test classes

Limraj commented on test/com/serotonin/mango/vo/report/CalculateConsolidatedChartHeightTestException.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.List;
+
+@RunWith(Parameterized.class)
+public class CalculateConsolidatedChartHeightTestException {
```
CalculateConsolidatedChartHeightExceptionTest

Limraj commented on test/com/serotonin/mango/vo/report/CalculateConsolidatedChartHeightTest.java:
```diff
@@ -0,0 +1,217 @@
+
+
+                {createPointStatisticsList(Arrays.asList(
+                        "Copy of Copy of Copy of Copy of virtual_ - dp_1",
```
Correct the names to make the test data clear what they mean.

Patrykb0802 committed:
Changes in test's datasets

Patrykb0802 committed:
Changed datasets used for CalculateConsolidatedChartHeightTest.java to be more understandable

Patrykb0802 committed:
Changed datapoint names in test datasets so they correspond to length of whole name

Patrykb0802 committed:
- Changed CHARACTERS_PER_LINE_IN_CHART_LEGEND_NUMBER cause now we limit how long data source and data point names are, so app behaves a bit different, now it should be good.
- Tested on app and changed some parameters in CalculateConsolidatedChartHeightTest.java

Limraj commented on test/com/serotonin/mango/vo/report/CalculateConsolidatedChartHeightTest.java:
```diff
@@ -0,0 +1,263 @@
+
+    @Parameterized.Parameters(name = "{index}: calculateConsolidatedChartHeight(points: {0}, expected height: {1},pixels per line: {2} base height of chart: {3}, characters per line: {4})")
+    public static Collection<Object[]> data() {
+        String _0CharactersPointName = "";
```
A local variable should not be named after the _ character, in some sources this character is used to denote private fields and methods.
The name may look like this:
 chars0PointName

Limraj commented on test/com/serotonin/mango/vo/report/CalculateConsolidatedChartHeightTest.java:
```diff
@@ -0,0 +1,263 @@
+
+
+            {createPointStatisticsList(Arrays.asList(
+                "ds_1 - 100characters100characters100characters100characters100characters100characters100characters10",
```
Move point names to variables as above to avoid repeating strings.

Patrykb0802 committed:
Moved point names to variables in CalculateConsolidatedChartHeightTest.java for better readability
